node/src/restarter.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{Node, NodeStorage};
+use arc_swap::ArcSwap;
+use config::{Committee, Parameters};
+use crypto::traits::{KeyPair, Signer, VerifyingKey};
+use executor::{ExecutionState, ExecutorOutput};
+use futures::future::join_all;
+use network::{PrimaryToWorkerNetwork, WorkerToPrimaryNetwork};
+use prometheus::Registry;
+use std::{fmt::Debug, path::PathBuf, sync::Arc};
+use tokio::sync::mpsc::{Receiver, Sender};
+use types::{PrimaryWorkerMessage, ReconfigureNotification, WorkerPrimaryMessage};
+
+// Module to start a node (primary, workers and default consensus), keep it running, and restarting it
+/// every time the committee changes.
+pub struct NodeRestarter;
+
+impl NodeRestarter {
+    pub async fn watch<Keys, PublicKey, State>(
+        keypair: Keys,
+        committee: &Committee<PublicKey>,
+        storage_base_path: PathBuf,
+        execution_state: Arc<State>,
+        parameters: Parameters,
+        mut rx_reconfigure: Receiver<(Keys, Committee<PublicKey>)>,
+        tx_output: Sender<ExecutorOutput<State>>,
+    ) where
+        PublicKey: VerifyingKey,
+        State: ExecutionState + Send + Sync + 'static,
+        State::Outcome: Send + 'static,
+        State::Error: Debug,
+        Keys: KeyPair<PubKey = PublicKey> + Signer<PublicKey::Sig> + Send + 'static,
```

Comment:
On the one hand, I'm glad you came around to trait contraints. On the other hand, please open an issue for me to mop this up later (you should be able to grab the most-constrained trait, i.e. KeyPair, and obtain the rest through projection, i.e. it's easier to do
```rust
foo<Keys: KeyPair>(
    ...
    Bar<Keys::PubKey>
    ...
)
```
rather than 
```rust 
foo<Keys: KeyPair<PubKey = PublicKey> + Signer<PublicKey::Sig> + Send + 'static, PublicKey: VerifyingKey> (
    ...
   Bar<PublicKey>
    ...
)
```
Moreover, `KeyPair` includes the Signer constraint, which you should not need to repeat)

Reply:
I also though so but if I remove the Signer trait I get the following error:
```
--> node/src/restarter.rs:53:35
   |
53 |             let primary_handles = Node::spawn_primary(
   |                                   ^^^^^^^^^^^^^^^^^^^ expected crypto::traits::VerifyingKey::Sig, found crypto::traits::KeyPair::Sig
   |
   = note: expected associated type `<PublicKey as VerifyingKey>::Sig`
              found associated type `<Keys as KeyPair>::Sig`
    ```

Reply:
Fixed #579